database/src/interface.rs

```diff
@@ -0,0 +1,987 @@
+    ///
+    /// No parameters is required.
+	fn height(&'service self) -> Result<u64,DB_FAILURES> {
+		let ro_tx = self.db.tx().map_err(Into::into)?;
+		ro_tx.num_entries::<table::blockhash>().map(|n| n as u64)
+	}
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+	// ----------------------------------| Blocks |---------------------------------
+
+	/// `add_block` add the block and metadata to the db.
+    ///
+    /// In case of failures, a `DB_FAILURES`
+    ///
+    /// Parameters:
+    /// `blk`: is the block to be added
+	/// `txs`: is the collection of transactions related to this block
+    /// `block_weight`: is the weight of the block (data's total)
+    /// `long_term_block_weight`: is the long term weight of the block (data's total)
+    /// `cumulative_difficulty`: is the accumulated difficulty at this block.
+    /// `coins_generated` is the number of coins generated after this block.
+    fn add_block(&'service self, blk: Block, txs: Vec<monero::Transaction>, block_weight: u64, long_term_block_weight: u64, cumulative_difficulty: u128, coins_generated: u64) 
+	-> Result<(), DB_FAILURES> 
+	{
+
+		// *sanity*
+		if blk.tx_hashes.len() != txs.len() {
+			return Err(DB_FAILURES::Other("sanity : Inconsistent tx/hashed sizes"));
+		}
+
+		let blk_hash = blk.id();
+
+		// let parent_height = self.height()?;
+
+		let mut num_rct_outs = 0u64;
+		self.add_transaction(blk.miner_tx.clone())?;
+		
+		if blk.miner_tx.prefix.version.0 == 2 {
+			num_rct_outs += blk.miner_tx.prefix.outputs.len() as u64;
+		}
+
+		// let mut tx_hash = Hash::null();
+		for tx in txs.into_iter()/*.zip(0usize..)*/ {
+			// tx_hash = blk.tx_hashes[tx.1];
+			for out in tx.prefix.outputs.iter() {
+				if out.amount.0 == 0 {
+					num_rct_outs += 1;
+				}
+			}
+			self.add_transaction(tx/*.0*/)?;
+		}
+
+		let blk_metadata = BlockMetadata {
+    		timestamp: blk.header.timestamp.0,
+    		total_coins_generated: coins_generated,
+    		weight: block_weight,
+    		cumulative_difficulty,
+    		block_hash: blk_hash.into(),
+    		cum_rct: num_rct_outs,
+    		long_term_block_weight,
+		};
+
+		self.add_block_data(blk, blk_metadata)
+	}
+
+	/// `add_block_data` add the actual block's data and metadata to the db. Underlying function of `add_block`
+    ///
+    /// In case of failures, a `DB_FAILURES` will be return.
+    ///
+    /// Parameters:
+    /// `blk`: is the block to add
+    /// `blk_metadata`: is the block's metadata to add
+    fn add_block_data(&'service self, blk: Block, mut blk_metadata: BlockMetadata) 
+	-> Result<(), DB_FAILURES>
+	{
+		let height = self.height()?;
+
+		let mut cursor_blockhash = self.write_cursor_dup::<table::blockhash>()?;
+		let mut cursor_blockmetadata = self.write_cursor_dup::<table::blockmetadata>()?;
+
+		if cursor_blockmetadata.get_dup(&(), &height)?.is_some() {
+			return Err(DB_FAILURES::AlreadyExist("Attempting to insert a block alreayd existent in the database"))?
+		}
```

Comment:
I think the height function is the issue. I'm returning the number of entries not the last block height. I should substract one to the result and it would corret everything

Reply:
Alright I'll fix it. I though the blockchain height was the top blocks height

Reply:
Replaced the check with the blockhash table.